File: swift-bridge/Sources/CoreImageBridge/FilterShape.swift
import CoreGraphics
import CoreImage
import Foundation

@_cdecl("ci_filter_shape_new")
public func ci_filter_shape_new(
    _ x: Double,
    _ y: Double,
    _ width: Double,
    _ height: Double
) -> UnsafeMutableRawPointer? {
    ci_retain(CIFilterShape(rect: CGRect(x: x, y: y, width: width, height: height)))
}

@_cdecl("ci_filter_shape_transform")
public func ci_filter_shape_transform(
    _ handle: UnsafeMutableRawPointer?,
    _ a: Double,
    _ b: Double,
    _ c: Double,
    _ d: Double,
    _ tx: Double,
    _ ty: Double,
    _ interior: Bool
) -> UnsafeMutableRawPointer? {
    guard let shape: CIFilterShape = ci_borrow(handle) else { return nil }
    let transform = CGAffineTransform(a: a, b: b, c: c, d: d, tx: tx, ty: ty)
    return ci_retain(shape.transform(by: transform, interior: interior))
}

@_cdecl("ci_filter_shape_inset")
public func ci_filter_shape_inset(
    _ handle: UnsafeMutableRawPointer?,
    _ dx: Int32,
    _ dy: Int32
) -> UnsafeMutableRawPointer? {
    guard let shape: CIFilterShape = ci_borrow(handle) else { return nil }
    return ci_retain(shape.insetBy(x: dx, y: dy))
}

@_cdecl("ci_filter_shape_union")
public func ci_filter_shape_union(
    _ handle: UnsafeMutableRawPointer?,
    _ otherHandle: UnsafeMutableRawPointer?
) -> UnsafeMutableRawPointer? {
    guard let shape: CIFilterShape = ci_borrow(handle),
          let other: CIFilterShape = ci_borrow(otherHandle)
    else {
        return nil
    }
    return ci_retain(shape.union(with: other))
}

@_cdecl("ci_filter_shape_union_rect")
public func ci_filter_shape_union_rect(
    _ handle: UnsafeMutableRawPointer?,
    _ x: Double,
    _ y: Double,
    _ width: Double,
    _ height: Double
) -> UnsafeMutableRawPointer? {
    guard let shape: CIFilterShape = ci_borrow(handle) else { return nil }
    return ci_retain(shape.union(with: CGRect(x: x, y: y, width: width, height: height)))
}

@_cdecl("ci_filter_shape_intersect")
public func ci_filter_shape_intersect(
    _ handle: UnsafeMutableRawPointer?,
    _ otherHandle: UnsafeMutableRawPointer?
) -> UnsafeMutableRawPointer? {
    guard let shape: CIFilterShape = ci_borrow(handle),
          let other: CIFilterShape = ci_borrow(otherHandle)
    else {
        return nil
    }
    return ci_retain(shape.intersect(with: other))
}

@_cdecl("ci_filter_shape_intersect_rect")
public func ci_filter_shape_intersect_rect(
    _ handle: UnsafeMutableRawPointer?,
    _ x: Double,
    _ y: Double,
    _ width: Double,
    _ height: Double
) -> UnsafeMutableRawPointer? {
    guard let shape: CIFilterShape = ci_borrow(handle) else { return nil }
    return ci_retain(shape.intersect(with: CGRect(x: x, y: y, width: width, height: height)))
}

@_cdecl("ci_filter_shape_extent")
public func ci_filter_shape_extent(
    _ handle: UnsafeMutableRawPointer?,
    _ outX: UnsafeMutablePointer<Double>?,
    _ outY: UnsafeMutablePointer<Double>?,
    _ outWidth: UnsafeMutablePointer<Double>?,
    _ outHeight: UnsafeMutablePointer<Double>?
) {
    guard let shape: CIFilterShape = ci_borrow(handle) else {
        outX?.pointee = 0
        outY?.pointee = 0
        outWidth?.pointee = 0
        outHeight?.pointee = 0
        return
    }
    let rect = shape.extent
    outX?.pointee = rect.origin.x
    outY?.pointee = rect.origin.y
    outWidth?.pointee = rect.size.width
    outHeight?.pointee = rect.size.height
}

File: src/filter_shape.rs
use core::ffi::c_void;
use core::fmt;
use core::ptr;

use apple_cf::cg::{CGAffineTransform, CGRect};

use crate::ffi;

/// A Core Image region shape used by accumulators and render plumbing.
pub struct CIFilterShape {
    ptr: *mut c_void,
}

impl Drop for CIFilterShape {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { ffi::ci_object_release(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

impl Clone for CIFilterShape {
    fn clone(&self) -> Self {
        Self {
            ptr: unsafe { ffi::ci_object_retain(self.ptr) },
        }
    }
}

impl fmt::Debug for CIFilterShape {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("CIFilterShape")
            .field("ptr", &self.ptr)
            .field("extent", &self.extent())
            .finish_non_exhaustive()
    }
}

impl CIFilterShape {
    pub(crate) const unsafe fn from_raw(ptr: *mut c_void) -> Self {
        Self { ptr }
    }

    fn from_non_null(ptr: *mut c_void, kind: &str) -> Self {
        assert!(!ptr.is_null(), "{kind} returned nil");
        unsafe { Self::from_raw(ptr) }
    }

/// Mirrors the `CoreImage` framework constant `fn`.
    pub const fn as_ptr(&self) -> *mut c_void {
        self.ptr
    }

/// Calls the `CoreImage` framework counterpart for `new`.
    pub fn new(rect: CGRect) -> Self {
        Self::from_non_null(
            unsafe { ffi::ci_filter_shape_new(rect.origin.x, rect.origin.y, rect.size.width, rect.size.height) },
            "CIFilterShape(rect:)",
        )
    }

/// Calls the `CoreImage` framework counterpart for `extent`.
    pub fn extent(&self) -> CGRect {
        let mut x = 0.0;
        let mut y = 0.0;
        let mut width = 0.0;
        let mut height = 0.0;
        unsafe { ffi::ci_filter_shape_extent(self.ptr, &mut x, &mut y, &mut width, &mut height) };
        CGRect::new(x, y, width, height)
    }

/// Calls the `CoreImage` framework counterpart for `transform`.
    pub fn transform(&self, transform: CGAffineTransform, interior: bool) -> Self {
        Self::from_non_null(
            unsafe {
                ffi::ci_filter_shape_transform(
                    self.ptr,
                    transform.a,
                    transform.b,
                    transform.c,
                    transform.d,
                    transform.tx,
                    transform.ty,
                    interior,
                )
            },
            "CIFilterShape.transform(by:interior:)",
        )
    }

/// Calls the `CoreImage` framework counterpart for `inset`.
    pub fn inset(&self, dx: i32, dy: i32) -> Self {
        Self::from_non_null(
            unsafe { ffi::ci_filter_shape_inset(self.ptr, dx, dy) },
            "CIFilterShape.insetBy(x:y:)",
        )
    }

/// Calls the `CoreImage` framework counterpart for `union`.
    pub fn union(&self, other: &Self) -> Self {
        Self::from_non_null(
            unsafe { ffi::ci_filter_shape_union(self.ptr, other.ptr) },
            "CIFilterShape.union(with:)",
        )
    }

/// Calls the `CoreImage` framework counterpart for `union_rect`.
    pub fn union_rect(&self, rect: CGRect) -> Self {
        Self::from_non_null(
            unsafe {
                ffi::ci_filter_shape_union_rect(self.ptr, rect.origin.x, rect.origin.y, rect.size.width, rect.size.height)
            },
            "CIFilterShape.union(with:)",
        )
    }

/// Calls the `CoreImage` framework counterpart for `intersection`.
    pub fn intersection(&self, other: &Self) -> Self {
        Self::from_non_null(
            unsafe { ffi::ci_filter_shape_intersect(self.ptr, other.ptr) },
            "CIFilterShape.intersect(with:)",
        )
    }

/// Calls the `CoreImage` framework counterpart for `intersection_rect`.
    pub fn intersection_rect(&self, rect: CGRect) -> Self {
        Self::from_non_null(
            unsafe {
                ffi::ci_filter_shape_intersect_rect(
                    self.ptr,
                    rect.origin.x,
                    rect.origin.y,
                    rect.size.width,
                    rect.size.height,
                )
            },
            "CIFilterShape.intersect(with:)",
        )
    }
}
